src/NuGet.Core/NuGet.Protocol.Core.v3/PackageUpdateResource.cs

```diff
@@ -0,0 +1,601 @@
+                    packagePath = packagePath + Path.DirectorySeparatorChar + '*';
+                }
+                else if (!packagePath.EndsWith("*", StringComparison.OrdinalIgnoreCase))
+                {
+                    packagePath = packagePath + '*';
+                }
+                packagePath = packagePath + NuGetConstants.PackageExtension;
+            }
+            return packagePath;
+        }
+
+        private static void EnsurePackageFileExists(string packagePath, IEnumerable<string> packagesToPush)
+        {
+            if (!packagesToPush.Any())
+            {
+                throw new ArgumentException(String.Format(CultureInfo.CurrentCulture,
+                    Strings.UnableToFindFile,
+                    packagePath));
+            }
+        }
+
+        // Indicates whether the specified source is a file source, such as: \\a\b, c:\temp, etc.
+        private static bool IsFileSource(string source)
+        {
+            Uri uri;
+            if (Uri.TryCreate(source, UriKind.RelativeOrAbsolute, out uri))
+            {
+                return uri.IsFile;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        // Pushes a package to the Http server.
+        private async Task PushPackageToServer(string source,
+            string apiKey,
+            string pathToPackage,
+            long packageSzie,
+            ILogger logger,
+            CancellationToken token)
+        {
+            HttpRequestMessage request = null;
+            HttpResponseMessage response = null;
+
+            try
+            {
+                response = await _httpSource.SendAsync(
+                    () => { return request = CreateRequest(request, source, pathToPackage, apiKey); },
+                    token);
+            }
+            finally
+            {
+                if (request != null)
+                {
+                    request.Dispose();
+                }
+            };
+
+            response.EnsureSuccessStatusCode();
+        }
+
+        private HttpRequestMessage CreateRequest(HttpRequestMessage currentRequest,
+            string source,
+            string pathToPackage,
+            string apiKey)
+        {
+            if (currentRequest != null)
+            {
+                //this should dispose the content, the file stream underneath, and everything.
+                currentRequest.Dispose();
+            }
+
+            var fileStream = new FileStream(pathToPackage, FileMode.Open, FileAccess.Read, FileShare.Read);
+            var request = new HttpRequestMessage(HttpMethod.Put, GetServiceEndpointUrl(source, string.Empty));
+            var content = new MultipartFormDataContent();
+
+            var packageContent = new StreamContent(fileStream);
+            packageContent.Headers.ContentType = MediaTypeHeaderValue.Parse("application/octet-stream");
+            content.Add(packageContent, "package", "package.nupkg");
```

Comment:
it it just a random field name used in the content body for later deserializing. I add comment for the randomness.